Sorted test results

cboos opened the pull request:
Now the meat of it ;-)

Minitest has a strong stance on randomizing the order of execution of tests.
That's all fine for the robustness of the testing itself, but when it comes to reading the report of the tests, this randomness can be very frustrating at times.

So the idea is to only produce some minimal output while the reporting progresses, and to make a more useful report at the end, first sorting the contexts, then within each context, sorting the tests themselves. You can then see at a glance where the failures and errors are located, and this alone can often be telling about the root cause.

I didn't sort the last FAILURES and ERRORS sections, because, well, you'd have to fix all of them anyway, right? So the order there doesn't matter as much.

Finally, there are some changes in the "fanciness" of the output: the colors are the more standard green/red/yellow now because the ones with the weights (or amber_300!) didn't work on Windows. To make up for that, I added some nice Unicode characters ;-)

cboos committed:
Present the results in condensed then sorted form.

In the first phase, we only show full details for the failed
tests, and print a '.' otherwise.

In the second phase, we show the full details for all the results,
 but we sort them by 'rank', which is either the expectation
number or the raw test name.

cboos committed:
Flush stdout/stderr before reporting for a test

This way, we make sure that the output for that test doesn't
come after we report about it, in the console.

eugeniobruno committed:
Fix instantiation of Results::Test

eugeniobruno committed:
Replace @rank attribute with polymorphic #sort_key methods

cboos commented:
Ah yes, well spotted for the Results::Test... I actually don't have any of those left in my code ;-)

And nice touch for the #sort_key, it's indeed better & simpler that way!

eugeniobruno committed:
Separate magic comments from code with a blank line

eugeniobruno committed:
Minor rename and delegation

eugeniobruno committed:
Change icons that occupy too much empty space

eugeniobruno committed:
Use fallback colors only when running on Windows

cboos commented on lib/minitest-bender/colorizer.rb:
```diff
@@ -0,0 +1,39 @@
+      end
+
+      def windows?
+        @windows ||= RbConfig::CONFIG['host_os'] =~ /mswin/
```
For me: `RbConfig::CONFIG['host_os'] # => "mingw32"`
```suggestion
        @windows ||= RbConfig::CONFIG['host_os'] =~ /mswin|mingw32/
```

eugeniobruno replied:
Fair enough. I figured the only troublemaker was the classic command prompt, but I guess not.

eugeniobruno committed:
Add mingw32 as a host for fallback colors

Co-authored-by: Christian Boos <cboos@bct-technology.com>

eugeniobruno commented:
The sorted results are great, and the unicode icons look cool. I definitely want all of that. However, I would still like to keep the old randomized list around too. Let me explain why.

There are gems and applications whose test suites run for several minutes, maybe even hours. Using the current version of the reporter, if I run one of those suites and I spot a failure or an error, I see which specific test it is, and its error message. Then, I can start to track and fix that issue without cancelling the suite execution.

This is of course the exception rather than the rule. Because of this, your new sorted style should be the default, and the old one should be opt-in. I'm thinking we could use the same strategy that is now in place to activate the reporter: A test option in TESTOPTS. For example, TESTOPTS=--bender-without-sorting.

What do you think?

cboos committed:
Merge branch 'master' into ranked-results

cboos commented:
(wanted to try out the "resolve conflicts" feature of GitHub)

So as for the TESTOPTS, yes, sure. But maybe even for the "long running" scenario, you might benefit from the sorted view at the end, so maybe instead of preventing the sorting to have the immediate results right away, maybe the flag could just enable the immediate more verbose output? (`--bender-immediate` or `--benber-verbose`?).

eugeniobruno commented:
Absolutely! Your proposal makes a lot of sense. I'm not sold on the name of the option, though. I'm leaning towards "--bender-realtime-details". But we can adjust that later on.

I'm not sure how much time I can dedicate to this during the week. Needless to say, feel free to try out an implementation.

eugeniobruno committed:
Minor code style improvements

eugeniobruno committed:
Restore the verbose recording style as opt-in

eugeniobruno commented:
It turns out it was simpler than I thought. I ended up choosing "--bender-verbose" for the option ;)

Unless you have further suggestions, I'm ready to merge this. It's awesome. Thanks a lot!

cboos committed:
Add two more options to control the output

With --bender-no-sorted-overview, the new section showing
the sorted tests by context and by order within a context
is simply omitted.

With --bender-verbose, the compact progress is replaced
with the normal detailed output, with an eye on long running
test suites where it is important to see the failure as soon
as they happen. In this spirit, we also directly show the
details of the failure.

cboos commented:
Oh, I missed your last two commits. In the meantime, I did something quite similar (also a `--bender-verbose` one ;-) 

See the last commit in ranked-results-alt branch (b8e7b55). You also may want to take bits of that or take it wholesale. I didn't try to merge it here so that you can more easily compare the two approaches. I can do the merge once we discussed it.

cboos committed:
Merge branch 'ranked-results-alt' into ranked-results

cboos committed:
Simplify report generation by having more stateful Result.

Instead of recomputing the State corresponding to one Result,
we just make the Result#state accessor public.

We also store the "execution count" in the Result.
This number corresponds to the number of results
of the same state created so far.

For example, the first Failing result has 1, second
Failing result has 2, then first Raising result has 1,
and so on.

cboos committed:
Use unicode separator between context and name

cboos committed:
Make location in error details compatible with Emacs

If that last ':' is not there, Emacs won't detect the line as an error location.

cboos commented:
If you don't mind me abusing this pull request for fixing one more unrelated minor glitch... I have to keep my Emacs happy!

eugeniobruno commented:
Sure, go right ahead. I will most likely review your latest commits sometime this weekend.

eugeniobruno committed:
Improved the handling of options

eugeniobruno committed:
Minor style improvement that was lost in a merge

eugeniobruno committed:
Freeze string literals in one more file

cboos commented:
That's all fine with me, btw. If you'd like to merge this one as it stands, please go ahead.

I have two more things "pending":
- a minor fix for another backtrace, but that's so unrelated so I'll create another pull request when time permits (tomorrow probably)
- I have one more idea for improving the look of the sorted results, but I don't know when I'll be able to get around to it, so that will be yet another pull request, later on

eugeniobruno commented:
Excellent, thank you!
